src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala

```diff
@@ -581,28 +585,41 @@ trait AppAndGroupFormats {
             (__ \ "acceptedResourceRoles").readNullable[Set[String]](nonEmpty) ~
             (__ \ "ipAddress").readNullable[IpAddress] ~
             (__ \ "version").readNullable[Timestamp].withDefault(Timestamp.now()) ~
-            (__ \ "residency").readNullable[Residency]
-          ) (ExtraFields)
+            (__ \ "residency").readNullable[Residency] ~
+            (__ \ "portDefinitions").readNullable[Seq[PortDefinition]]
+          )(ExtraFields)
             .filter(ValidationError("You cannot specify both uris and fetch fields")) { extra =>
               !(extra.uris.nonEmpty && extra.fetch.nonEmpty)
             }
             .filter(ValidationError("You cannot specify both an IP address and ports")) { extra =>
-              extra.maybePorts.forall(_.isEmpty) || extra.ipAddress.isEmpty
+              val appWithoutPorts = extra.maybePorts.forall(_.isEmpty) && extra.maybePortDefinitions.forall(_.isEmpty)
+              appWithoutPorts || extra.ipAddress.isEmpty
+            }
+            .filter(ValidationError("You cannot specify both ports and port definitions")) { extra =>
+              extra.maybePorts.isEmpty || extra.maybePortDefinitions.isEmpty
```

Comment:
As you already suggested: if ports and portDefinitions hold the same data, we could allow this.